### ISSUE TITLE
Error configuring property: camel.component.test[00001]port[80]tcp.addr because cannot find component with name test[00001]port[80]tcp.

### ISSUE THREAD

**Issue by lsergio** [apache/camel-k#5179]
### What happened?

I have created the following Integration:
```
apiVersion: camel.apache.org/v1
kind: Integration
metadata:
  name: basic
spec:
  sources:
  - name: flow1.groovy
    content: |-
      from('quartz://flow1/trigger?cron=0+*+*+?+*+*')
         .to("https://httpbin.org/delay/2?bridgeEndpoint=true")
         .to("log:info")
```

It creates a CronJob to run every minute. 
When the Integration runs, I get the following error on its log:

```
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Error configuring property: camel.component.test[00001]port[80]tcp.addr because cannot find component with name test[00001]port[80]tcp. Make sure you have the component on the classpath
	at org.apache.camel.quarkus.main.CamelMainRuntime.start(CamelMainRuntime.java:65)
	at org.apache.camel.quarkus.core.CamelBootstrapRecorder.start(CamelBootstrapRecorder.java:45)
	... 15 more
Caused by: java.lang.IllegalArgumentException: Error configuring property: camel.component.test[00001]port[80]tcp.addr because cannot find component with name test[00001]port[80]tcp. Make sure you have the component on the classpath
	at org.apache.camel.main.BaseMainSupport.lambda$doAutoConfigurationFromProperties$29(BaseMainSupport.java:1683)
	at org.apache.camel.main.MainHelper.computeProperties(MainHelper.java:407)
	at org.apache.camel.main.BaseMainSupport.doAutoConfigurationFromProperties(BaseMainSupport.java:1675)
	at org.apache.camel.main.BaseMainSupport.autoConfigurationFromProperties(BaseMainSupport.java:1638)
	at org.apache.camel.main.BaseMainSupport.autoconfigure(BaseMainSupport.java:447)
	at org.apache.camel.main.MainSupport.autoconfigure(MainSupport.java:73)
	at org.apache.camel.main.BaseMainSupport.postProcessCamelContext(BaseMainSupport.java:627)
	at org.apache.camel.quarkus.main.CamelMain.initCamelContext(CamelMain.java:116)
	at org.apache.camel.quarkus.main.CamelMain.doInit(CamelMain.java:86)
	at org.apache.camel.support.service.BaseService.init(BaseService.java:78)
	at org.apache.camel.quarkus.main.CamelMain.startEngine(CamelMain.java:137)
	at org.apache.camel.quarkus.main.CamelMainRuntime.start(CamelMainRuntime.java:49)
	... 16 more
```
00001 seems to be a knative revision number, so I decided to disable the knative trait:

```
  traits:
    knative:
      enabled: false
```
and the error disappeared.

As this integration did not create a Service/Knative Service, should the knative trait be activated?

This started to happen today in our cluster, so maybe it's also related to some environment configuration. If someone can provide some guidance, I would appreciate as well.

### Camel K version

2.2.0

**Comment by squakez** [comment]
Thanks for reporting. Is this happening only for cron based Integrations, or does it appear also on plain deployments?

**Comment by lsergio** [comment]
@squakez Here is an update with a few other scenarios:

**Scenario 1**
Quartz integration, Cron Trait configured with fallback:true
```
spec:
  sources:
  - name: flow1.groovy
    content: |-
      from('quartz://flow1/trigger?cron=0+*+*+?+*+*')
         .to("https://httpbin.org/delay/2?bridgeEndpoint=true")
         .to("log:info")
  traits:
    cron:
      fallback: true
```
**Results**
* Camel K created a deployment and not a CronJob (as expected)
* No Service or Knative Service/Revision was created (as expected)
* The deployment pod failed with the reported error
* Disabling the knative trait fixed the error

**Scenario 2**
Rest Integration, knatice-service trait disabled so the service is internal to the cluster only
```
spec:
  sources:
  - name: flow1.groovy
    content: |-
      from('rest:GET:/demo')
         .to("https://httpbin.org/delay/2?bridgeEndpoint=true")
         .to("log:info")
  traits:
    knative-service:
      enabled: false
```
**Results**
* Camel K created a deployment and a ClusterIP Service (as expected)
* No Knative Service/Revision was created (as expected)
* The deployment pod failed with the reported error
* Disabling the knative trait fixed the error

**Scenario 3**
Rest Integration, knatice-service trait enabled so the service is exposed with Knative
```
spec:
  sources:
  - name: flow1.groovy
    content: |-
      from('rest:GET:/demo')
         .to("https://httpbin.org/delay/2?bridgeEndpoint=true")
         .to("log:info")
  traits:
    knative-service:
      enabled: true
      minScale: 1
```
**Results**
* Camel K created a Knative Service (as expected)
* Knative created 2 services: `basic-00001` and `basic-00001-private`
* The deployment pod failed with the reported error
* Knative revision status remained with Ready = Unknown and Reason = Deploying
* Disabling the knative trait fixed the error, and it changed the number of services Knative created: now there are `basic`, `basic-0001` and `basic-00001-private` services.


So now I'm wondering if the Knative trait should play a role in scenarios where a service is not created - scenarios 1 & 2. And trying to understand why knative didn't create the 3 services in the 3rd scenario.

**Comment by lsergio** [comment]
Comparing the Knative services before and after the fix in the 3rd scenario, I see that when knative trait is enabled, the Knative Service declares the following environment variable:
```
        - name: CAMEL_KNATIVE_CONFIGURATION
          value: '{"services":[]}'
```
Per the docs, this env var is expected, but I'm not sure whether the service list should be empty.

**Comment by lsergio** [comment]
After further investigation into knative objects, I see the private endpoint was not assigned IP addresses:
```
basic-00001                                                   10.251.165.73:8012,10.251.165.73:8112   4m21s
basic-00001-private                                                                                   4m21s
```
And describing the private one shows me:
```
Name:         basic-00001-private
Namespace:    a1b2c3d4e5
Labels:       app=basic-00001
              camel.apache.org/integration=basic
              networking.internal.knative.dev/serverlessservice=basic-00001
              networking.internal.knative.dev/serviceType=Private
              serving.knative.dev/configuration=basic
              serving.knative.dev/configurationGeneration=1
              serving.knative.dev/configurationUID=4cbbd31b-a712-4e3f-a524-af181bf50662
              serving.knative.dev/revision=basic-00001
              serving.knative.dev/revisionUID=4156d7bc-7743-4ef9-a78c-6f4a57179372
              serving.knative.dev/service=basic
              serving.knative.dev/serviceUID=8ed0cca8-fe97-44c8-850e-c93e392acd49
Annotations:  endpoints.kubernetes.io/last-change-trigger-time: 2024-02-26T13:49:25Z
Subsets:
  Addresses:          <none>
  NotReadyAddresses:  10.251.159.135
  Ports:
    Name             Port  Protocol
    ----             ----  --------
    http-usermetric  9091  TCP
    http             8012  TCP
    http-queueadm    8022  TCP
    http-istio       8012  TCP
    https            8112  TCP
    http-autometric  9090  TCP

Events:  <none>
```
The ip address is marked as NotReady. Looking into that...

**Comment by lsergio** [comment]
In order to perform an isolated knative test, I exported the knative service created by the integration and modified it to create a new ksvc:
- changed the image to ghcr.io/knative/helloworld-go:latest
- removed the mounted volumes
- removed the heatlh check probes
The final result:
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  annotations:
    serving.knative.dev/creator: kubernetes-admin
    serving.knative.dev/lastModifier: system:serviceaccount:a1b2c3d4e5:camel-k-operator
  labels:
    bindings.knative.dev/include: "true"
    camel.apache.org/generation: "1"
    camel.apache.org/integration: basic
  name: basic
  namespace: a1b2c3d4e5
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/minScale: "1"
      creationTimestamp: null
      labels:
        camel.apache.org/integration: basic
    spec:
      containerConcurrency: 0
      containers:
      - env:
        - name: TARGET
          value: "Go Sample v1"      
        image: ghcr.io/knative/helloworld-go:latest
        name: integration
      enableServiceLinks: false
      imagePullSecrets:
      - name: gcr-json-key
      timeoutSeconds: 300
  traffic:
  - latestRevision: true
    percent: 100
```

This ksvc worked all 3 services were created.
```
k get svc | grep basic
basic                                                         ClusterIP   None             <none>        80/TCP                                               4m9s
basic-00001                                                   ClusterIP   172.20.70.175    <none>        80/TCP,443/TCP                                       4m10s
basic-00001-private                                           ClusterIP   172.20.1.209     <none>        80/TCP,443/TCP,9090/TCP,9091/TCP,8022/TCP,8012/TCP   4m10s
```
Then I changed the ksvc and added a heath probe that would never succeed:
```
        readinessProbe:
          successThreshold: 1
          tcpSocket:
            port: 3000  
```
And now I get the same situation where the pod fails, and the basic service is not created.

So I think I can assume that the lack of one svc is actually a consequence, and not a cause of the pod failure.

**Comment by lsergio** [comment]
If knative trait is disabled, the list of properties is shorter:
```
2024-02-26 16:43:22,900 DEBUG [org.apa.cam.mai.BaseMainSupport] (main) Properties from Camel properties component:
2024-02-26 16:43:22,900 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.service.port.metrics=8080
2024-02-26 16:43:22,900 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.port[8080]tcp.addr=172.20.253.1
2024-02-26 16:43:22,900 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.mount.path.configmaps=/etc/camel/conf.d/_configmaps
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.integration=basic
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.sources[0].name=flow1
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.port[8080]tcp=tcp://172.20.253.1:8080
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.version=2.2.0
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.kubernetesConfig.clientEnabled=false
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.service.port=8080
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.port[8080]tcp.proto=tcp
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.conf=/etc/camel/application.properties
2024-02-26 16:43:22,901 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.digest=vR4kTZd4vpDiwffwgyUmRWbb7fgbEboiC-nNvy5qvCYM
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.service.host=172.20.253.1
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.sources[0].location=file:/etc/camel/sources/flow1.groovy
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.port[8080]tcp.port=8080
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.runtime.version=3.2.3
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.mount.path.secrets=/etc/camel/conf.d/_secrets
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.conf.d=/etc/camel/conf.d
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.metrics.port=tcp://172.20.253.1:8080
2024-02-26 16:43:22,902 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.operator.id=camel-k
2024-02-26 16:43:22,903 DEBUG [org.apa.cam.mai.BaseMainSupport] (main)     camel.k.sources[0].language=groovy
```

**Comment by lsergio** [comment]
After further investigation, I found out that another namespace in the same cluster that has Camel-K installed with the same configuration didn't reproduce the error.
So, for testing, I decided do delete the integration kits from the failing namespace and reapplied the integration. And then it worked.

So, at the end, I think I could not provide enough information to fix the issue, so maybe we might just close it.

**Comment by lburgazzoli** [comment]
can you check the env vars available in the pod ?

**Comment by lsergio** [comment]
@squakez I am using operator 2.2.0, camel-k-runtime 3.2.2.
I could not find these values in any configuration either. I agree, it's really weird, specially considering that it was fixed after deleting and rebuilding the integration kits.

@lburgazzoli I can't reproduce the issue anymore, but I compared the ksvc's with and without the knative trait enabled, and the only differences were the CAMEL_KNATIVE_CONFIGURATION environment var, with no services declared in it. And also the list of jar files used in the arguments.

**Comment by lburgazzoli** [comment]
Things like `private.service.port.http.usermetric` seems to be something injected i.e. by prometheus or whatnot and may be something that it is injected as en var. Camel does look also at the env var when trying to auto configure stuffs so it may be possible that some weird naming ends up being processed by camel as a property.

So it would have been very useful to see the POD's env vars to confirm the hypothesis

**Comment by lburgazzoli** [comment]
can you get the env var of the running pod ?

**Comment by lsergio** [comment]
@lburgazzoli Isn't it the list in the spec included in the previous message?
If you mean connect to the container and get the vars, I'm having trouble doing so, as the containes fails to start:

```
k exec -it demo008-00002-deployment-57447bf67-cpjsz -- c integration env
Defaulted container "integration" out of: integration, queue-proxy
error: unable to upgrade connection: container not found ("integration")
```

**Comment by lburgazzoli** [comment]
I need the actual env var of the pod as some env var are injected by k8s

**Comment by lsergio** [comment]
Here it is:

```
root@demo008-00001-deployment-7687bd4f74-792jk:/deployments# ls -R /etc/camel
/etc/camel:
application.properties  sources

/etc/camel/sources:
flow1.groovy


root@demo008-00001-deployment-7687bd4f74-792jk:/deployments# cat /etc/camel/application.properties 
camel.k.sources[0].language = groovy
camel.k.sources[0].location = file:/etc/camel/sources/flow1.groovy
camel.k.sources[0].name = flow1

root@demo008-00001-deployment-7687bd4f74-792jk:/deployments# cat /etc/camel/sources/flow1.groovy 
from('rest:GET:/hello1') .to("https://httpbin.org/delay/2?bridgeEndpoint=true") .to("log:info")
from('rest:GET:/hello2') .to("log:debug")
```

**Comment by lburgazzoli** [comment]
even more weird, is there a way to replicate it ?

**Comment by lsergio** [comment]
I don't think so. Even in another namespace within the same cluster it works. I'm trying strategies like removing the build kits, cleaning the maven repo, restarting the node.
It seems to be random, but might be triggered by adding or removing some jar file to the classpath depending on what components or traits I use.

**Comment by lsergio** [comment]
One interesting test:

I cleared the integration kits, and created 2 copies of the same integration.
One of them used the routine build strategy, the other one used pod build strategy.

Only the one that used the routine strategy failed. I'll take a look at the jar files within the operator maven repo to try and file something wrong.

**Comment by lburgazzoli** [comment]
can you also grab the env var of the operator ?

**Comment by lsergio** [comment]
Bingo, we found it:

```
I have no name!@camel-k-operator-74f8c76949-t5pgb:/$ env
CAMEL_COMPONENT_TEST_00002_PRIVATE_PORT_8022_TCP_PROTO=tcp

(there are 300+ env vars here, most of them similar to the ones above)
```

I don't see these variables in the deployment, so somehow k8s injected them. Let me show this to the person who created the cluster and installed Camel K.

**Comment by lburgazzoli** [comment]
O guess some of them could be marked as 'build time runtime fixed' by quakus, hence they may get persisted in the runner jar

**Comment by lsergio** [comment]
@lburgazzoli After reaching to some people here, I found the root cause...

Kubernetes is injecting environment variables into any pod with metadata based on the services that are running, as in:

```
APIM_CONNECTOR_C2228101_CF6A_4B3E_B502_62C99359A799_SERVICE_PORT=tcp://172.20.203.161:80
APIM_CONNECTOR_08BFC757_3D93_419A_8182_390B77D62DB1_SERVICE_PORT_80_TCP=tcp://172.20.206.40:80
APIM_CONNECTOR_08BFC757_3D93_419A_8182_390B77D62DB1_SERVICE_PORT_80_TCP_ADDR=172.20.206.40
```
That's not enough to explain the CAMEL_COMPONENT variables.
But then we found that a developer was also testing Camel components using Camel K and created an integration named "camel-component-test".
This created a camel-component-test service that was active when the operator started, and thus explains the weird env vars.

When I tested using the pod strategy it didn't fail because the the camel-component-test service is no longer running. 

So now everything is running fine again and I advised everyone about not using camel-* names.

Thanks a lot for the guidance, @lburgazzoli. 
I think we can close the ticket now and I'll remove the bug label.
